java/core/src/main/java/sleeper/core/statestore/AllFileReferences.java

```diff
@@ -16,61 +16,67 @@
  */
 public class AllFileReferences {
+    private final Set<String> filesWithNoReferences;
+    private final Set<FileInfo> activeFiles;
+    private final boolean moreThanMax;
 
-    private final Set<FileReferences> files;
+    public AllFileReferences(Collection<FileInfo> activeFiles, Collection<String> filesWithNoReferences) {
+        this(activeFiles, filesWithNoReferences, false);
+    }
 
-    public AllFileReferences(Set<FileReferences> files) {
-        this.files = files;
+    public AllFileReferences(Collection<FileInfo> activeFiles, Collection<String> filesWithNoReferences, boolean moreThanMax) {
+        this.filesWithNoReferences = new LinkedHashSet<>(filesWithNoReferences);
+        this.activeFiles = new LinkedHashSet<>(activeFiles);
+        this.moreThanMax = moreThanMax;
     }
 
-    public static AllFileReferences fromActiveFilesAndReferenceCounts(
+    public static AllFileReferences fromActiveFilesAndReadyForGCFiles(
             Stream<FileInfo> activeFiles,
-            Stream<FileReferenceCount> fileReferenceCounts) {
-        Map<String, List<FileInfo>> referencesByFilename = new LinkedHashMap<>();
-        Map<String, FileReferenceCount> referenceCountByFilename = fileReferenceCounts
-                .peek(counts -> referencesByFilename.put(counts.getFilename(), new ArrayList<>()))
-                .collect(Collectors.toMap(FileReferenceCount::getFilename, Function.identity()));
-        activeFiles.forEach(file -> referencesByFilename
-                .computeIfAbsent(file.getFilename(), name -> new ArrayList<>())
-                .add(file));
-        Set<FileReferences> fileReferences = referencesByFilename.entrySet().stream()
-                .map(entry -> new FileReferences(entry.getKey(), referenceCountByFilename.get(entry.getKey()).getLastUpdateTime(), entry.getValue()))
-                .collect(Collectors.toUnmodifiableSet());
-        return new AllFileReferences(fileReferences);
+            Stream<String> filesWithNoReferences,
+            boolean moreFilesWithNoReferences) {
+        return new AllFileReferences(
+                activeFiles.collect(Collectors.toList()),
+                filesWithNoReferences.collect(Collectors.toList()),
+                moreFilesWithNoReferences);
+    }
+
+    public static AllFileReferences fromStateStoreWithReadyForGCLimit(StateStore stateStore, int maxFilenamesReadyForGC) throws StateStoreException {
+        Iterator<String> filenamesReadyForGC = stateStore.getReadyForGCFilenamesBefore(Instant.ofEpochMilli(Long.MAX_VALUE)).iterator();
+        List<String> readyForGC = new ArrayList<>();
+        int count = 0;
+        while (filenamesReadyForGC.hasNext() && count < maxFilenamesReadyForGC) {
+            readyForGC.add(filenamesReadyForGC.next());
+            count++;
+        }
+        boolean moreThanMax = filenamesReadyForGC.hasNext();
+        return new AllFileReferences(stateStore.getActiveFiles(), readyForGC, moreThanMax);
```

Comment:
This doesn't produce a consistent snapshot. This could end up with files being missed out of the report if they were marked ready for GC after the ready for GC files are loaded from the state store.